### ISSUE TITLE
Roadmap: SmolLM2 Inference (RoPE + RMSNorm + SwiGLU + GQA)

### ISSUE THREAD

**Issue by ziliangpeng** [ziliangpeng/triton-llm#19]
# SmolLM2 — Triton Inference Engine

## Goal
Implement SmolLM2-135M / 360M / 1.7B inference in pure Python + Triton (no PyTorch), using the exact same kernel infrastructure as GPT-2.

SmolLM2 is 100% Llama-architecture compatible: RoPE + RMSNorm + SwiGLU + GQA + SentencePiece.

## Architecture vs GPT-2

| Component | GPT-2 (done) | SmolLM2 (new) | What changes? |
|-----------|-------------|---------------|---------------|
| Position Encoding | Learned PE (embedding.py) | **RoPE** | New kernel needed |
| Normalization | LayerNorm (layernorm.py) | **RMSNorm** | New kernel (simpler) |
| Activation | GELU (gelu.py) | **SwiGLU** | New kernel (gate × activation) |
| Attention | MHA (attention.py) | **GQA** | attention.py rewrite: KV group broadcast + RoPE |
| Config | GPT2Config (config.py) | LlamaConfig | New config file |
| Model | GPT2Model (model.py) | LlamaModel | New model wiring |

## SmolLM2-135M Config (from HF)
```json
{
  "hidden_size": 576,       "num_hidden_layers": 30,
  "num_attention_heads": 9, "num_key_value_heads": 3,
  "intermediate_size": 1536, "vocab_size": 49152,
  "max_position_embeddings": 8192,
  "rms_norm_eps": 1e-5,     "rope_theta": 100000,
  "tie_word_embeddings": true,
  "hidden_act": "silu",     "torch_dtype": "bfloat16"
}
```

## Why SmolLM2 (not Llama directly)
- **Apache 2.0 license** — fully open
- **No gated access** — download directly from HF, no forms
- **135M is GPT-2 Small size** — fast iteration on same dev pod
- **"LlamaForCausalLM"** in architectures — kernel-level compatible with real Llama
- **8K context** — KV cache will show dramatic speedup

## Why multi-agent
Each PR is a single atomic component. The multi-agent dev loop (Implementer → Reviewer → Tester → Gatekeeper) catches bugs GPU tests miss.

### 🔲 P0 — Infrastructure + Config
- [ ] Create `smollm2_triton/` package structure
- [ ] `smollm2_triton/config.py` — SmolLM2 variants (135M/360M/1.7B)
- [ ] Weights download script (`scripts/download_smollm2_weights.py`)
- [ ] Reference: `gpt2_triton/config.py`, `gpt2_triton/gpu.py` (reuse)

### 🔲 P1 — RMSNorm Kernel
- [ ] `smollm2_triton/kernels/rms_norm.py` — RMSNorm Triton kernel
- [ ] Tests: correctness vs numpy reference, edge cases
- [ ] Reference: `gpt2_triton/kernels/layernorm.py` pattern

### 🔲 P2 — RoPE Kernel
- [ ] `smollm2_triton/kernels/rope.py` — Rotary Position Embedding
- [ ] Precompute cos/sin table, apply in-place on Q and K
- [ ] Tests: numerical match vs reference, non-power-of-2 head_dim

### 🔲 P3 — SwiGLU Kernel
- [ ] `smollm2_triton/kernels/swiglu.py` — SiLU(x) * x_gate
- [ ] Tests: vs numpy `silu(x) * gate`, empty, edge cases

### 🔲 P4 — GQA + RoPE Attention Kernel
- [ ] Extend `attention.py` with GQA (group broadcast) + RoPE (apply inside kernel)
- [ ] Tests: GQA correctness (9→3 heads), RoPE correctness, KV cache compatibility
- [ ] Edge: non-power-of-2 head_dim (576/9=64 — fine), 8K sequence

### 🔲 P5 — SmolLM2 Model Wiring
- [ ] `smollm2_triton/model.py` — full LlamaForCausalLM inference
- [ ] `smollm2_triton/tokenizer.py` — SentencePiece wrapper
- [ ] `scripts/run_smollm2.py` — end-to-end demo
- [ ] 8K+ KV cache benchmark

### 🔲 P6 — Real Weight Verification
- [ ] Download SmolLM2-135M weights (no gate 🎉)
- [ ] Compare logits vs HF reference
- [ ] Greedy generation quality check
- [ ] Run on 360M, 1.7B

### 🔲 Deferred
- FP16 / mixed precision support
- Pre-allocated KV cache (PagedAttention)
- Batch > 1

**Comment by ziliangpeng** [comment]
### P4 ✅ GQA + RoPE Attention (PR #23)

- GQA attention kernel with causal/non-causal, absolute-position mask, 2D grid
- Pointer hoisting,  as 
- 7 GQA tests (equivalence, reduced KV, RoPE+e2e, causal, causal-seq_k>seq_q, non-causal, empty)
- Gemini review: 4 rounds, converged
- 78/78 regression pass on H100 (gcp5)

On to P5 + P6!
